packages/parser/src/language/sankey/sankeyValueConverter.ts

```diff
@@ -0,0 +1,42 @@
+import type { CstNode, GrammarAST, ValueType } from 'langium';
+
+import { MermaidValueConverter } from '../common/valueConverter.js';
+import { sankeyLinkNodeRegex, sankeyLinkValueRegex } from './sankeyMatcher.js';
+
+const rulesRegexes: Record<string, RegExp> = {
+  SANKEY_LINK_NODE: sankeyLinkNodeRegex,
+  SANKEY_LINK_VALUE: sankeyLinkValueRegex,
+};
+
+export class SankeyValueConverter extends MermaidValueConverter {
+  protected runCustomConverter(
+    rule: GrammarAST.AbstractRule,
+    input: string,
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    _cstNode: CstNode
+  ): ValueType | undefined {
+    const regex: RegExp | undefined = rulesRegexes[rule.name];
+    if (regex === undefined) {
+      return undefined;
+    }
+    const match = regex.exec(input);
+    if (match === null) {
+      return undefined;
+    }
+    // source and target with double quote and value
+    if (match[1] !== undefined) {
+      if (rule.name === 'SANKEY_LINK_VALUE') {
+        return Number(match[1].replaceAll('"', ''));
```

Comment:
Forget about features and differences of jison and languim for a moment, and consider a grammar, imagine it's kinda in pure form:

```
field -> escaped | non_escaped
escaped -> " escaped_text "
non_escaped -> non_escaped_text
```

But this line says, that we are making quotes to be parts of values and then extracting them. That's sort of double work. If we are extracting number the lexer should return NUMBER (capital letters mean it is the token for tokenization step) with the value alongside it (it can be a string, which we convert later on to number, but without quotes). In our case grammar for value should be like:

```
value -> " number " | number
```

The same can be applied to node (source or target) name. Quotes must not be a part of node. This case is more difficult though, because of quotes escapement.

Reply:
Langium team actually use this approach:

<https://github.com/eclipse-langium/langium/blob/f3cf20687c8a891935bb3425f9d12242aeae37ab/packages/langium/src/grammar/langium-grammar.langium#L210-L210>

---

In some rules, like this one, where we need to override `TokenBuilder` and use a custom pattern function:

<https://github.com/mermaid-js/mermaid/blob/5fe951a6925818fe4d0c2b8c6b15a7849b86cece/packages/parser/src/language/sankey/sankeyTokenBuilder.ts#L13-L17>

It's possible to do as you suggested, since this rule can match everything except `"` in normal mode. But would I recommend it? idk.

But in other rules, e.g., pie section label:

<https://github.com/mermaid-js/mermaid/blob/8c31db63521b8726a80275d0267ca68c66823970/packages/parser/src/language/pie/pie.langium#L18>

It must be wrapped between doable quotes. In this case, it won't be recommended to create a rule that matches everything except `"` and then override it in `TokenBuilder`.